Issue85and86

brettviren opened the pull request:
This addresses the two related issues #85 and #86 

- copy all top level .jsonnet files to `$PREFIX/share/wirecell/`.
- move `funcs.jsonnet` content to `pgraph.jsonnet` and leave behind backward compatibility code.
- if `--install-config=<name>` given then install also the `pgrapher/common/` files and the configuration and data files for experiment `<name>`.

This PR also adds `wire-cell-data` as a git submodule so that `wcb` build can find the files.  This submodule is only required for this purpose.  Developers or users that get their data in other ways may leave the submodule dangling.

The data files are identified based on the `.files` data object value from the experiment's `params.jsonnet` file.  If data files are named outside this structure they will not be picked up by the install method.

brettviren commented:
Not wrong, just different.

Waf options are not implicitly saved into the waf cache.  An option given to `wcb configure` does not automatically persist to other commands.  Any persistence of information requires some waf code to explicitly add some values into `cfg.env` inside `configure()`.  We could do this but in this case I figured we may want to rerun this install command multiple times, once for each experiment.  Eg:

```
./wcb configure --prefix=`pwd`/install [...]
./wcb --notests install
cd install && tar -cf /tmp/wirecell-main.tar . && cd -
./wcb install --install-config=pdsp
cd install && tar -cf /tmp/wirecell-pdsp.tar share/wirecell && cd -
rm -rf install/share/wirecell
./wcb install --install-config=icarus
cd install && tar -cf /tmp/wirecell-icarus.tar share/wirecell && cd -
```
I think this sequence is okay if maybe a bit awkward.  We could add more `wscript` code to make `wcb` create these tarfiles more automatically.

One issue that may need rethinking: as-is, each of these experiment tar files will include the top-level `*.jsonnet` files which will also be included in the main tar file (ie, that which might turn into a `wirecell` UPS product).  I think this will be okay and may even be helpful:

* it means experiments need to set `WIRECELL_PATH` to just one directory, the one provided by their `wirecell-<experiment>` UPS product (or something else if the experiment tar file is unpacked in an existing experiment config product area).
* the top `*.jsonnet` files are relatively stable but if an experiment must update them faster than we make a future release this gives them an avenue.

We can certainly do this stuff differently so let me know if you have suggestions.

HaiwangYu commented:
Thanks @brettviren , I now understand how to use this.
Previously I was trying to use the method proposed in #86 .


> One issue that may need rethinking: as-is, each of these experiment tar files will include the top-level *.jsonnet files which will also be included in the main tar file (ie, that which might turn into a wirecell UPS product). I think this will be okay and may even be helpful:

I think this procedure is good to use, maybe need to confirm with @wenqiang-gu 

Then I have two technical points:
 - This [commit](https://github.com/WireCell/wire-cell-data/commit/73dcc0dffa3bac88af87fb475c2296a41e39e3c2) is needed, otherwise, the building would fail when `--install-config` is not specified
 - This [line](https://github.com/WireCell/wire-cell-data/blob/master/wscript_build#L27) seems need to be changed to `if one: dat_files.add(str(one))`, otherwise, building would fail in my python 2.7 environment.

brettviren commented:
Thanks for these checks, @HaiwangYu 

The string issue is a Python2 vs Python3 difference.  I'm glad you are testing with Python2.  Your fix is also okay for Python3.

I added your string fix to `wire-cell-data` and updated this repo's submodule link to pick them up.

brettviren commented:
@HaiwangYu thanks for catching the blunder.  Latest appended commit fixes the problem where omitting the `--install-config` would cause ALL experiments' config to install.  Default still includes the `common` parts.

wenqiang-gu commented:
Thanks @HaiwangYu , the configurations (*jsonnet, *fcl) are installed correctly under `/wcdo/lib/ups/wirecell/wctdev/Linux64bit+4.15-2.17-sl7-7-e19-prof/share/wirecell`. However, I don't see the data files installed. Is it expected if I run `./wcb -p --notests --install-config=icarus install`?

brettviren commented:
My guess is that your `data/` submodule is not populated.  

Try first:

```
git submodule init
git submodule update
```

And then repeat your `install`.

We should discuss if this should remain optional or be required.  My feeling is many types of developers will not want the extra copy of data around and that only someone making UPS or other official releases really require it and they can assure it is available in their release scripts.  But, I'm open to enacting some other policy.

wenqiang-gu commented:
> @wenqiang-gu My guess is that your `data/` submodule is not populated.
> 
> Try first:
> 
> ```
> git submodule init
> git submodule update
> ```

Thanks. This works fine now.
